### PR TITLE
Fix "srpski" not finding results in Wikipedia compact links

### PR DESCRIPTION
**Pull request by Nikerabbit**
See also https://phabricator.wikimedia.org/T178996

ULS compact links passes list of languages that have redirects resolved, hence "sr" appears as "sr-cyrl". The search api returns "sr" and it didn't match.

### CODE REVIEW

**Comment by amire80** [comment]
Looks like an improvement to me, although it's not totally perfect. It still doesn't find "srpski" on Wikipedia, for example, although it does seem to resolve the issues with Punjabi, Azeri, and Kazakh, which are listed at https://phabricator.wikimedia.org/T178996 .

**Comment by Nikerabbit** [comment]
I'm pretty sure it will find srpski, although it is impossible for me to test on Wikipedia. How did you test?

**Comment by amire80** [comment]
Here's how I tested:
* Go to https://ca.wikipedia.org/wiki/Portada?debug=true (it's important for Punjabi, because https://gerrit.wikimedia.org/r/#/c/386158/ is already deployed there)
* Open the jquery.uls.languagefilter.js in the debugger.
* Change the code to the code in this patch.

The code in this patch helps find Punjabi (pa) with "punj", Kazakh with "kaz" and "каз", and Azerbaijani with "azer". It doesn't find "srpski" (sr) with "srp"; it only finds srpslohrvatski (sh).

**Comment by Nikerabbit** [comment]
LanguageNameSearchData.php has
`                        'srpski' => 'sr',`

The api should return this result (you can inspect the HTTP response separately) and with this patch the client code should match it to sr-cyrl (that actually goes to 'sr' in the end because of overrides).

**Comment by amire80** [comment]
Hmm... The idea is fairly simple:
- In the context of (compact) interlanguage links, all the languages in which the article is available must be shown, and all of them must be findable. In particular, those that are listed under "Variants and redirects" in https://phabricator.wikimedia.org/T178996
- If I type "srp" in the search box, and article is available in both "Srpski" and "Srpskohrvatski", I expect to see both languages in the search results.
- If I search for "срп", I also want to be able to find "Srpski". (In Wikipedia we have Srpskohrvatski only in the Latin script, so it's OK if we don't find it in this case. It would be nice if it's found, too, but not a requirement.)
- Only find each language once. Don't show "Srpski" more than once.
- I know it sounds too generic, but don't break other things :)

**Comment by Pl217** [comment]
Query "срп" ("ср" and "с" as well) gives two "Српски / srpski" results. Here is the [screenshot](https://phabricator.wikimedia.org/F13139157).

> If I search for "срп", I also want to be able to find "Srpski". In Wikipedia we have Srpskohrvatski only in the Latin script, so it's OK if we don't find it in this case. It would be nice if it's found, too, but not a requirement.

We have **srpskohrvatski** in results for "срп" query. List of found languages (in code) has this structure:
```
{
    sh: "српскохрватски",
    sr: "српски / srpski",
    sr-ec: "српски (ћирилица)"
}
```
Both `sr` and `sr-ec` are redirected to `sr-cyrl` and we have same result twice. For "srp" search query, `sr-el` redirects to `sr-latn` which is not found in list of languages (data structure in code) and we don't have duplication, but we might, in theory.

---

Here are other search queries from "Languages with script variants and redirects" section on [T178996](https://phabricator.wikimedia.org/T178996). Checked on dewiki, which doesn't have latest changes deployed, but relevant code should be same.
- **каз** - placeholder _каза́хскїй_ shown and **Қазақша** found in results (with kk language code).
- **az**, **azer** - placeholder _azerbaidjaneg_ shown and **Azərbaycanca** (az language code) and **تۆرکجه** (azb language code) found, but South Azerbaijani displayed twice. That happens for same reasons explained above for Serbian. `az`, `az-arab` and `azb` languages are found for "az" and "azer" query, and `azb` redirects to `az-arab` so South Azerbaijani variant is displayed twice.
- **azerb**, **azerba** - This finds `az` and `azb` language codes, and both **Azərbaycanca** and **تۆرکجه** are shown once.
- **аз**, **азер** - placeholder _азербайджанийн_ shown and only **Azərbaycanca** found. This one is probably expected and definitely not related to redirects and variants.
- **pun** - placeholder _punjabi_ shown and both ਪੰਜਾਬੀ (Punjabi, pa language code) and پنجابی (Western Punjabi, pnb language code) found.

**Comment by Nikerabbit** [comment]
So the duplication would be fine if both `sr` and `sr-cyrl` was requested (so it would actually be two different codes but with identical name). But since only one of them was requested, we should filter it out.

**Comment by Pl217** [comment]
Latest commit deals successfully with previously raised problems of duplicate languages, so I have removed "Changes requested" review.
I would like small code changes, will make comments inline.

File: src/jquery.uls.languagefilter.js [hunk 1]
```diff
@@ -207,15 +207,35 @@
 				var autofillLabel,
 					results = [];
 
-				$.each( result.languagesearch, function ( code, name ) {
-					if ( this.options.languages[ code ] ) {
+				$.each( result.languagesearch, function ( apiCode, name ) {
+					var code,
+						redirect = $.uls.data.isRedirect( apiCode );
```
Comment (Pl217) [inline src/jquery.uls.languagefilter.js]: I don't like the fact we're checking for redirect every single one of language search results and doing that unconditionally, but will not make that a blocker. I would just like slightly reorganized code.

Comment (Nikerabbit) [reply src/jquery.uls.languagefilter.js]: It is a very cheap call, and this structure allows to keep the logic simple. Please show how would you do it better :)

Comment (Pl217) [reply src/jquery.uls.languagefilter.js]: Yes, I agree that call is cheap. Achieving same, simple logic, with slightly rearranged structure is more appealing to me. Here is one way of doing it:

```
var code, redirect;

if ( this.options.languages[ apiCode ] ) {
    code = apiCode;
} else {
    redirect = $.uls.data.isRedirect( apiCode );

    if ( !redirect || !this.options.languages[ redirect ] ) {
        return;
    }

    code = redirect;
}
```

**Comment by Pl217** [comment]
Although problems like [T142971](https://phabricator.wikimedia.org/T142971) might be solved similarly to how duplicates are removed in this patch, that is not a blocker to this change.

All languages examined in my previous comments are easily found.

**Pull request merged**